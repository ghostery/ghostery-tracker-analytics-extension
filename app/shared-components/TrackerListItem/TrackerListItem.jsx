import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';
import OverlayTrigger from 'react-bootstrap/OverlayTrigger';
import Tooltip from 'react-bootstrap/Tooltip';

import TrackerTimelineGraph from '../TrackerTimelineGraph';
import { truncateString } from '../../../src/utils/convert';
import copyText from '../../utils/javascript/copyText';
import './TrackerListItem.scss';

class TrackerListItem extends Component {
  constructor(props) {
    super(props);
    this.state = { hoveredRequest: '' };
  }

  componentDidMount() {
    const { actions, id } = this.props;
    actions.addTrackerListItemRef(id, this.selfRef);
  }

  handleHover(key) {
    this.setState({ hoveredRequest: key });
  }

  copyToClipboard(fullUrl) {
    const { actions, showToasts } = this.props;
    copyText(fullUrl);
    if (showToasts) { actions.openToast({ toastText: 'URL copied!' }, 2500); }
  }

  render() {
    const {
      id,
      name,
      size,
      latency,
      latencyCategory,
      requests,
      showGraph,
      handleExpand,
      expanded,
      firstTracker,
      starred,
      handleStar,
      highlighted,
      onMouseEnter,
      onMouseLeave,
      tooltipRef,
      pageEventLinesHeight,
    } = this.props;
    const { hoveredRequest } = this.state;
    const trackerListItemClasses = ClassNames(
      'TrackerListItem', 'd-flex',
      { expanded, highlighted },
    );
    const trackerListItemInformaton = ClassNames(
      'TrackerListItem__trackerInformation',
      { showGraph },
    );
    const trackerListItemHeader = ClassNames(
      'TrackerListItem__header',
      'd-flex',
      'align-items-center',
      'justify-content-start',
      { showGraph },
    );
    const trackerListItemStar = ClassNames(
      'TrackerListItem__star',
      { small: !showGraph },
    );
    const trackerListItemCircle = ClassNames(
      'TrackerListItem__circle',
      'd-flex',
      'align-items-center',
      'justify-content-center',
      latencyCategory,
      { small: !showGraph },
    );
    const trackerListItemStats = ClassNames(
      'TrackerListItem__stats',
      { small: !showGraph },
    );
    const trackerListItemStatsInfo = ClassNames(
      'TrackerListItem__statsInfo',
      'd-flex',
      'justify-content-start',
      { small: !showGraph },
    );
    const trackerListItemTrackerSize = ClassNames(
      'TrackerListItem__trackerSize',
      { small: !showGraph },
    );
    const trackerListItemRequestList = ClassNames(
      'TrackerListItem__requestList',
      { expanded, small: !showGraph },
    );
    const trackerListItemRequest = ClassNames(
      'TrackerListItem__request',
      'd-flex',
      'align-items-center',
      'justify-content-start',
      { small: !showGraph, empty: requests.length === 0 },
    );
    const tooltipClass = ClassNames(
      'tooltipLeft',
      { tooltip__panelTrackerList: !showGraph },
    );
    const starIconFilePath = starred
      ? 'dist/images/shared/star-filled.svg'
      : 'dist/images/shared/star-empty.svg';

    return (
      <div
        className={trackerListItemClasses}
        onMouseEnter={() => onMouseEnter(id)}
        onMouseLeave={() => onMouseLeave(id)}
        ref={(node) => { this.selfRef = node; }}
      >
        <div className={trackerListItemInformaton}>
          <div className={trackerListItemHeader} onClick={() => handleExpand(id)}>
            <img
              alt="Star Tracker"
              src={[chrome.extension.getURL(starIconFilePath)]}
              className={trackerListItemStar}
              onClick={(e) => {
                e.stopPropagation();
                handleStar(id, name);
              }}
            />
            <div className={trackerListItemCircle}>
              <p className="TrackerListItem__circleText">{requests.length}</p>
            </div>
            <div className={trackerListItemStats}>
              <p className="TrackerListItem__statsHeader">{name}</p>
              <div className={trackerListItemStatsInfo}>
                <p className={trackerListItemTrackerSize}>{`Size: ${size}`}</p>
                <p>{`Latency: ${latency}`}</p>
              </div>
            </div>
          </div>
          <div className={trackerListItemRequestList}>
            {requests.length > 0 ? (
              requests.map((request) => {
                const trackerListRequestName = ClassNames(
                  'TrackerListItem__requestName',
                  'd-flex',
                  'justify-content-start',
                  'align-items-center',
                  { small: !showGraph, hovered: (hoveredRequest === request.requestId) },
                );
                return (
                  <div
                    key={`tracker:${id} requestStart:${request.startTime}`}
                    className={trackerListItemRequest}
                  >
                    <OverlayTrigger
                      placement={showGraph ? 'right' : 'top'}
                      container={tooltipRef}
                      overlay={(
                        <Tooltip className={tooltipClass}>
                          {showGraph ? request.src : 'Click to copy'}
                        </Tooltip>
                      )}
                    >
                      <div
                        className={trackerListRequestName}
                        onMouseEnter={() => this.handleHover(request.requestId)}
                        onMouseLeave={() => this.handleHover('')}
                        onClick={() => this.copyToClipboard(request.src)}
                      >
                        <img
                          src={[chrome.extension.getURL('dist/images/shared/copy-icon.svg')]}
                          width="10"
                          height="13"
                          alt="More info hover: Average data usage by all trackers detected on the current page"
                        />
                        <p>
                          {showGraph
                            ? truncateString(request.src, 32)
                            : truncateString(request.src, 36)}
                        </p>
                      </div>
                    </OverlayTrigger>
                  </div>
                );
              })) : (
                <div className={trackerListItemRequest}>
                  <p>No requests were completed for this tracker</p>
                </div>
            )}
          </div>
        </div>
        {showGraph && (
          <TrackerTimelineGraph
            requests={requests}
            expanded={expanded}
            firstTracker={firstTracker}
            pageEventLinesHeight={pageEventLinesHeight}
            latencyCategory={latencyCategory}
            handleClick={() => handleExpand(id)}
          />
        )}
      </div>
    );
  }
}

TrackerListItem.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  size: PropTypes.string.isRequired,
  latency: PropTypes.string.isRequired,
  latencyCategory: PropTypes.string.isRequired,
  requests: PropTypes.arrayOf(PropTypes.shape({
    src: PropTypes.string.isRequired,
    endTime: PropTypes.number.isRequired,
    startTime: PropTypes.number.isRequired,
  })).isRequired,
  showGraph: PropTypes.bool.isRequired,
  handleExpand: PropTypes.func.isRequired,
  expanded: PropTypes.bool.isRequired,
  firstTracker: PropTypes.bool.isRequired,
  pageEventLinesHeight: PropTypes.number.isRequired,
  starred: PropTypes.bool.isRequired,
  handleStar: PropTypes.func.isRequired,
  highlighted: PropTypes.bool.isRequired,
  onMouseEnter: PropTypes.func.isRequired,
  onMouseLeave: PropTypes.func.isRequired,
  showToasts: PropTypes.bool.isRequired,
  actions: PropTypes.shape({
    addTrackerListItemRef: PropTypes.func.isRequired,
    openToast: PropTypes.func.isRequired,
  }).isRequired,
  tooltipRef: PropTypes.instanceOf(PropTypes.element),
};

TrackerListItem.defaultProps = {
  tooltipRef: null,
};

export default TrackerListItem;
