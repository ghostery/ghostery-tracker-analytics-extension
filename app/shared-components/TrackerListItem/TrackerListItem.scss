@import '../../utils/scss/color-variables.scss';

.TrackerListItem {
  margin: 1px 2px;
  border: 1px solid transparent;

  &.expanded {
    padding-bottom: 10px;
  }

  &.highlighted {
    border: 1px solid $ghosty-blue;
  }

  &:first-of-type {
    margin-top: 0;
  }
}

.TrackerListItem__trackerInformation {
  width: 250px;

  &.showGraph {
    margin-right: 3px;
    min-width: 300px;
  }
}

.TrackerListItem__header {
  background-color: $opaque-background-dark;
  height: 50px;
  cursor: pointer;

  &.showGraph {
    height: 62px;
  }
}

.TrackerListItem__star {
  position: relative;
  top: -17px;
  left: 7px;
  margin-right: 10px;
  width: 13px;
  height: 13px;
  cursor: pointer;

  &.small {
    top: -13px;
  }
}

.TrackerListItem__circle {
  margin: 0 18px 0 22px;
  border-radius: 28px;
  width: 28px;
  height: 28px;

  &.green {
    border: 3px solid $info-green;
  }

  &.yellow {
    border: 3px solid $info-yellow;
  }

  &.red {
    border: 3px solid $info-red;
  }

  &.small {
    margin: 0 15px 0 12px;
  }
}

.TrackerListItem__circleText {
  font-size: 15.8px;
}

.TrackerListItem__stats {
  width: 190px;
  font-size: 11px;

  &.small {
    width: 160px;
    font-size: 9px;
  }

  span {
    height: 25px;
  }
}

.TrackerListItem__statsHeader {
  margin: 0;
  font-weight: 600;
}

.TrackerListItem__statsInfo {
  p {
    margin: 9px 0 0;
  }

  &.small > p {
    margin: 6px 0 0;
  }
}

.TrackerListItem__trackerSize {
  width: 90px;

  &.small {
    width: 75px;
  }
}

.TrackerListItem__requestList {
  display: none;

  &.expanded {
    display: inline;
  }

  &.small {
    padding-top: 4px;
  }
}

.TrackerListItem__request {
  padding-top: 1px;
  font-size: 11px;
  color: $ghosty-blue;

  &.empty {
    padding-left: 51px;
  }

  &.small {
    padding-top: 0;
    font-size: 9px;
    line-height: 1px;
  }

  &.small:first-of-type {
    margin-top: 5px;
  }
}

.TrackerListItem__requestName {
  margin-left: 97px;
  cursor: pointer;

  > img {
    margin-right: 7px;
    display: none;
  }

  &.small {
    margin-left: 51px;
  }

  &.small > p {
    margin: 5px;
  }

  &.hovered {
    margin-left: 77px;
    font-weight: bold;

    > img {
      display: block;
    }

    &.small {
      margin-left: 31px;
    }
  }
}
