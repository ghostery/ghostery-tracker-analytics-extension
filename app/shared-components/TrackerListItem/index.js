import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import TrackerListItem from './TrackerListItem.jsx';
import { addTrackerListItemRef, openToast } from '../../store/actions/UserInterfaceActions';

const mapStateToProps = state => ({
  tooltipRef: state.UserInterfaceReducer.tooltipRef,
  showToasts: state.SettingsReducer.localSettings.showToasts,
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ addTrackerListItemRef, openToast }, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(TrackerListItem);
