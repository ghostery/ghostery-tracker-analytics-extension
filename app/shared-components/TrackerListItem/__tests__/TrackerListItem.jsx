import React from 'react';
import { shallow } from 'enzyme';
import chrome from 'sinon-chrome';
import path from 'path';
import fakeProps from 'react-fake-props';

import TrackerListItem from '../TrackerListItem';

jest.mock('../../../../src/classes/Globals', () => ({
  BROWSER_INFO: { displayName: 'Chrome' },
}));


describe('app/shared-components/TrackerListItem/', () => {
  global.chrome = chrome;
  describe('Snapshot test with shallow', () => {
    test('TrackerListItem should render', () => {
      const componentPath = path.join(__dirname, '../TrackerListItem.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<TrackerListItem {...props} />);
      expect(component).toMatchSnapshot();
    });
  });

  describe('Functionality testing', () => {
    test('handleHover should resove with the set state', () => {
      const componentPath = path.join(__dirname, '../TrackerListItem.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<TrackerListItem {...props} />);
      component.setState({ hoveredRequest: 'test' });

      expect(component.find('.TrackerListItem__requestName').length).toBe(1);
    });

    test('copyToClipboard should be called', () => {
      const componentPath = path.join(__dirname, '../TrackerListItem.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<TrackerListItem {...props} />);
      const instance = component.instance();
      const copyToClipboardMock = jest.spyOn(instance, 'copyToClipboard').mockImplementation(() => {});

      component.find('.TrackerListItem__requestName').simulate('click');
      expect(copyToClipboardMock.mock.calls.length).toBe(1);
    });
  });
});
