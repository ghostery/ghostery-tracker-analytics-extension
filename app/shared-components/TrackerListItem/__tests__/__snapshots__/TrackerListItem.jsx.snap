// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/shared-components/TrackerListItem/ Snapshot test with shallow TrackerListItem should render 1`] = `
<div
  className="TrackerListItem d-flex expanded highlighted"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="TrackerListItem__trackerInformation showGraph"
  >
    <div
      className="TrackerListItem__header d-flex align-items-center justify-content-start showGraph"
      onClick={[Function]}
    >
      <img
        alt="Star Tracker"
        className="TrackerListItem__star"
        onClick={[Function]}
        src={
          Array [
            undefined,
          ]
        }
      />
      <div
        className="TrackerListItem__circle d-flex align-items-center justify-content-center latencyCategory"
      >
        <p
          className="TrackerListItem__circleText"
        >
          1
        </p>
      </div>
      <div
        className="TrackerListItem__stats"
      >
        <p
          className="TrackerListItem__statsHeader"
        >
          name
        </p>
        <div
          className="TrackerListItem__statsInfo d-flex justify-content-start"
        >
          <p
            className="TrackerListItem__trackerSize"
          >
            Size: size
          </p>
          <p>
            Latency: latency
          </p>
        </div>
      </div>
    </div>
    <div
      className="TrackerListItem__requestList expanded"
    >
      <div
        className="TrackerListItem__request d-flex align-items-center justify-content-start"
        key="tracker:id requestStart:1"
      >
        <OverlayTrigger
          container={null}
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              className="tooltipLeft"
              placement="right"
            >
              requests.src
            </Tooltip>
          }
          placement="right"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <div
            className="TrackerListItem__requestName d-flex justify-content-start align-items-center"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <img
              alt="More info hover: Average data usage by all trackers detected on the current page"
              height="13"
              src={
                Array [
                  undefined,
                ]
              }
              width="10"
            />
            <p>
              requests.src
            </p>
          </div>
        </OverlayTrigger>
      </div>
    </div>
  </div>
  <Connect(TrackerTimelineGraph)
    expanded={true}
    firstTracker={true}
    handleClick={[Function]}
    latencyCategory="latencyCategory"
    pageEventLinesHeight={1}
    requests={
      Array [
        Object {
          "endTime": 1,
          "src": "requests.src",
          "startTime": 1,
        },
      ]
    }
  />
</div>
`;
